service/worker/scanner/README.md

```diff
@@ -0,0 +1,247 @@
+		ScannerWFTypeName: ConcreteExecutionsScannerWFTypeName, // also used as context key
+		FixerWFTypeName:   ConcreteExecutionsFixerWFTypeName,   // also used as context key
+		DynamicParams:     shardscanner.DynamicParams{...},     // contains enabled flags and other limits
+		DynamicCollection: dc,
+		ScannerHooks:      ConcreteExecutionHooks,       // constructor-func for this impl
+		FixerHooks:        ConcreteExecutionFixerHooks,  // constructor-func for this impl
+		...
+	}
+}
+```
+There's quite a large chunk of complicated code reuse, value-driven behavior rather than type-driven, and back and forth
+between changing based on arguments or hardcoded class-method-like lookups underneath all this, but the hooks themselves
+are relatively readable after only a couple layers of navigation.
+
+They broadly summarize as:
+- Config has hooks and dynamic config, some of which defines invariants.
+- Hooks have a "manager" which basically collects invariants.
+- Hooks have an "iterator" which yields a page of data each time it is called (the shared iterator handles forwarding paging tokens).
+- Some of all of the above behavior is achieved by using methods on the iota that defines the invariant
+  (which is often selected from serialized data by its string name).
+
+Some of these fields and stuff from the high-level shared `Resources` object are combined
+to build the `ShardScanner` or `ShardFixer` per shard, which is the thing that drives the "read -> process -> write" loops
+at the core of these processes:
+```go
+scanner := NewScanner(
+    shardID,
+    ctx.Hooks.Iterator(activityCtx, pr, params),
+    resources.GetBlobstoreClient(),
+    params.BlobstoreFlushThreshold,
+    ctx.Hooks.Manager(activityCtx, pr, params, resources.GetDomainCache()),
+    func() { activity.RecordHeartbeat(activityCtx, heartbeatDetails) },
+    scope,
+    resources.GetDomainCache(),
+)
+```
+
+And last but not least, note that:
+- Config is saved when the workflow starts, not read from scratch when the activities begin.
+- Scanner/fixer are on an every-minute cron schedule, but are allowed to run for ludicrous amounts of time.
+- This means any config changes will _very likely_ need to wait for the current run to finish before picking up your changes.
+  - You can advance this relatively quickly by terminating the existing workflows, and then restarting a workers-service.
+  - Terminating will stop the cron completely, but restarting any instance will start a new cron, which will pick up the new config.
+
+<details><summary>Enabling scanners for local dev</summary>
+
+For local purposes, you can pretty easily enable these via the dynamicconfig
+file store [config/dynamicconfig/development.yaml](../../../config/dynamicconfig/development.yaml):
+```yaml
+# inconsistent key names, unfortunately.  don't guess, copy/paste
+worker.concreteExecutionFixerEnabled:
+- value: true
+  constraints: {}
+worker.executionsScannerInvariantCollectionStale:
```

Comment:
Is the 'fixing' behaviour enabled by default?

Reply:
after a bit of digging: it appears both "yes" and "oh no".  I'm debating changing it...

the long and the short of it is that fixer does not have any per-invariant config.  what it does is:
- reads executions that scanner identified as corrupted
- ignores what it decided / what scan identified the problem
- runs all invariants on that execution (after checking that the domain is enabled)

So there is both no way to not-fix, and by introducing a new fix you start *immediately* using it on all scan results.

bleh.

I think I can get a scanner-config-like addition in fixer without much trouble, so we can toggle invariants for both scanner and fixer independently.  doing anything fancier might take a lot more work :\